halo2_proofs/src/circuit/value.rs

```diff
@@ -612,6 +612,45 @@ impl<F: Field> Mul<F> for Value<&Assigned<F>> {
     }
 }
 
+impl<F: PartialEq> PartialEq<Value<F>> for Value<F> {
+    fn eq(&self, other: &Value<F>) -> bool {
+        if let Some((a, b)) = self.inner.as_ref().zip(other.inner.as_ref()) {
+            a.eq(b)
+        } else {
+            self.inner.is_none() == other.inner.is_none()
+        }
+    }
+}
+impl<F: PartialEq> Eq for Value<F> {}
+
+impl<F: PartialOrd> PartialOrd<Value<F>> for Value<F> {
+    fn partial_cmp(&self, other: &Value<F>) -> Option<std::cmp::Ordering> {
+        match (self.inner.as_ref(), other.inner.as_ref()) {
+            (None, None) => Some(std::cmp::Ordering::Equal),
+            (None, Some(_)) => Some(std::cmp::Ordering::Less),
+            (Some(_), None) => Some(std::cmp::Ordering::Greater),
```

Comment:
Not sure we should say greater/less when we have Some/None.

Reply:
Hmmm so `Unknown` is less than `0`?

I'm not sure on which should be the behaviour sincerely. Feels to me that comparing `Some` to `None` should fail somehow as we can only compare things that are Some or things that are None. But not a mix.

Reply:
These are fair points @adria0.
Unsure on how to proceed now TBH.

Reply:
I think we can maybe merge this and see if any errors pop out related to that.

Reply:
I feel we shouldn't implement `PartialOrd` for `Value`, imagine this in synthesize:

```rust
let f = region.assign_fixed(...)?; // Value::unknown() when it's create_proof
let a = region.assign_advice(...)?; // Value::unknown() when it's keygen
if f.value() < a.value() {
    foo()
} else {
    bar()
}
```

Then this will not incur error but will make the `keygen` to do `bar` but `create_proof` to do `foo`.

Reply:
This seems fine to me!  I don't see any problem in defining that Unknown is the smallest value, and after it come the known values.  This allows sorting and comparing Values (in a stable way), which is the purpose of this implementation.

Reply:
I see, but I still feel this `PartialCmp` impl might make introduce some case that is confusing (the case above) since it leaks the information of `Value` which should be opaque when synthesizing.

> `Value` already mixes `Some` with `None` internally: `Value::known(x) + Value::unknown() == Value::unknown` is internally a `Some(x) + None  == None`

I think this is fine because the output is still `Value`, but the output of `PartialCmp` is not `Value` anymore.

> It's used here:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/5ba2cf199415c57356e03c0e3875a0a2ad1bdd44/zkevm-circuits/src/evm_circuit/execution.rs#L1214

In this case what if we have something like:

```diff
         let rlc_assignments: BTreeSet<_> = block
             .rws
             .table_assignments()
             .iter()
             .map(|rw| {
                 challenges
                     .evm_word()
                     .map(|randomness| rw.table_assignment_aux(randomness).rlc(randomness))
             })
-            .collect();
+            .fold(BTreeSet<_>::new(), |mut set, value| {
+                value.map(|value| set.insert(value));
+                set
+            });
```

Reply:
Thanks for updating!